File: 2389-longest-subsequence-with-limited-sum/2389-longest-subsequence-with-limited-sum.cs
public class Solution {
    public int[] AnswerQueries(int[] nums, int[] queries)
{
    Array.Sort(nums);
    int[] answer = new int[queries.Length];

    for (int i = 1; i < nums.Length; i++)
    {
        nums[i] += nums[i - 1];
    }

    for (int i = 0; i < queries.Length; i++)
    {
        answer[i] = binarySearchAnswerQueries(nums, queries[i]);
    }

    return answer;
}

public int binarySearchAnswerQueries(int[] nums, int val)
{
    int left = 0, right = nums.Length;

    while (left < right)
    {
        int mid = left + (right - left) / 2;

        if (val < nums[mid])
            right = mid;
        else
            left = mid + 1;
    }

    return left;
}
}

File: 2389-longest-subsequence-with-limited-sum/2389-longest-subsequence-with-limited-sum.cpp
class Solution {
public:
    int ansQuery_BinarySearch(vector<int>& nums, int val) {
    int left = 0, right = nums.size();

    while (left < right) {
        int mid = left + (right - left) / 2;

        if (nums[mid] > val)
            right = mid;
        else
            left = mid + 1;
    }

    return left;
}

vector<int> answerQueries(vector<int>& nums, vector<int>& queries) {
    sort(nums.begin(), nums.end());

    int n = nums.size();
    vector<int> prefix, ans;

    for (int i = 0; i < n; i++) {
        int val = nums[i];
        val += prefix.empty() ? 0 : prefix.back();
        prefix.push_back(val);
    }

    for (auto q : queries) {
        ans.push_back(ansQuery_BinarySearch(prefix, q));
    }

    return ans;
}
};